// poku/lib/modules/helpers/list-files.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.listFiles = exports.getAllFiles = exports.escapeRegExp = exports.isFile = exports.sanitizePath = void 0;
const node_process_1 = require("process");
const node_path_1 = require("path");
const fs_js_1 = require("../../polyfills/fs.js");
const files_js_1 = require("../../configs/files.js");
const regex = {
    sep: /[/\\]+/g,
    pathLevel: /(\.\.(\/|\\|$))+/g,
    unusualChars: /[<>|^?*]+/g,
    absolutePath: /^[/\\]/,
    safeRegExp: /[.*{}[\]\\]/g,
    defaultFilter: /\.(test|spec)\./i,
};
const sanitizePath = (input, ensureTarget) => {
    const sanitizedPath = input
        .replace(regex.sep, node_path_1.sep)
        .replace(regex.pathLevel, '')
        .replace(regex.unusualChars, '');
    return ensureTarget
        ? sanitizedPath.replace(regex.absolutePath, `.${node_path_1.sep}`)
        : sanitizedPath;
};
exports.sanitizePath = sanitizePath;
const isFile = async (fullPath) => (await (0, fs_js_1.stat)(fullPath)).isFile();
exports.isFile = isFile;
const escapeRegExp = (string) => string.replace(regex.safeRegExp, '\\$&');
exports.escapeRegExp = escapeRegExp;
const envFilter = ((_a = node_process_1.env.FILTER) === null || _a === void 0 ? void 0 : _a.trim())
    ? new RegExp((0, exports.escapeRegExp)(node_process_1.env.FILTER), 'i')
    : undefined;
const getAllFiles = async (dirPath, files = new Set(), configs) => {
    let isFullPath = false;
    const currentFiles = await (async () => {
        if (await (0, exports.isFile)(dirPath)) {
            isFullPath = true;
            return Array.prototype.concat((0, exports.sanitizePath)(dirPath));
        }
        return await (0, fs_js_1.readdir)((0, exports.sanitizePath)(dirPath));
    })();
    const filter = envFilter
        ? envFilter
        : (configs === null || configs === void 0 ? void 0 : configs.filter) instanceof RegExp
            ? configs.filter
            : regex.defaultFilter;
    const exclude = (configs === null || configs === void 0 ? void 0 : configs.exclude)
        ? Array.isArray(configs.exclude)
            ? configs.exclude
            : [configs.exclude]
        : undefined;
    await Promise.all(currentFiles.map(async (file) => {
        const fullPath = isFullPath ? dirPath : (0, node_path_1.join)(dirPath, file);
        const stat = await (0, fs_js_1.stat)(fullPath);
        if (fullPath.indexOf('node_modules') !== -1 ||
            fullPath.indexOf('.git/') !== -1) {
            return;
        }
        if (isFullPath && (files_js_1.states === null || files_js_1.states === void 0 ? void 0 : files_js_1.states.isSinglePath)) {
            return files.add(fullPath);
        }
        if (exclude) {
            for (const pattern of exclude) {
                if (pattern.test(fullPath)) {
                    return;
                }
            }
        }
        if (filter.test(fullPath)) {
            return files.add(fullPath);
        }
        if (stat.isDirectory()) {
            await (0, exports.getAllFiles)(fullPath, files, configs);
        }
    }));
    return files;
};
exports.getAllFiles = getAllFiles;
const listFiles = async (targetDir, configs) => Array.from(await (0, exports.getAllFiles)((0, exports.sanitizePath)(targetDir), new Set(), configs));
exports.listFiles = listFiles;
